// @shopify/shopify-app-react-router/dist/cjs/server/authenticate/admin/billing/request.js
'use strict';

var shopifyApi = require('@shopify/shopify-api');
require('../../../types.js');
var respondToInvalidSessionToken = require('../../helpers/respond-to-invalid-session-token.js');
var invalidateAccessToken = require('../../helpers/invalidate-access-token.js');
require('isbot');
var helpers = require('./helpers.js');

function requestBillingFactory(params, request, session) {
    return async function requestBilling({ plan, isTest, returnUrl, ...overrides }) {
        const { api, logger } = params;
        logger.info('Requesting billing', {
            shop: session.shop,
            plan,
            isTest,
            returnUrl,
        });
        let result;
        try {
            result = await api.billing.request({
                plan: plan,
                session,
                isTest,
                returnUrl,
                returnObject: true,
                ...overrides,
            });
        }
        catch (error) {
            if (error instanceof shopifyApi.HttpResponseError && error.response.code === 401) {
                logger.debug('API token was invalid, responding to invalid session', {
                    shop: session.shop,
                });
                await invalidateAccessToken.invalidateAccessToken(params, session);
                throw respondToInvalidSessionToken.respondToInvalidSessionToken({
                    params,
                    request,
                    retryRequest: true,
                });
            }
            throw error;
        }
        throw helpers.redirectOutOfApp(params, request, result.confirmationUrl, session.shop);
    };
}

exports.requestBillingFactory = requestBillingFactory;
