// @shopify/shopify-app-react-router/dist/esm/server/authenticate/admin/billing/cancel.mjs
import { HttpResponseError } from '@shopify/shopify-api';
import '../../../types.mjs';
import { respondToInvalidSessionToken } from '../../helpers/respond-to-invalid-session-token.mjs';
import { invalidateAccessToken } from '../../helpers/invalidate-access-token.mjs';
import 'isbot';

function cancelBillingFactory(params, request, session) {
    return async function cancelBilling(options) {
        const { api, logger } = params;
        logger.debug('Cancelling billing', { shop: session.shop, ...options });
        try {
            return await api.billing.cancel({
                session,
                subscriptionId: options.subscriptionId,
                isTest: options.isTest,
                prorate: options.prorate,
            });
        }
        catch (error) {
            if (error instanceof HttpResponseError && error.response.code === 401) {
                logger.debug('API token was invalid, responding to invalid session', {
                    shop: session.shop,
                });
                await invalidateAccessToken(params, session);
                throw respondToInvalidSessionToken({
                    params,
                    request,
                    retryRequest: true,
                });
            }
            throw error;
        }
    };
}

export { cancelBillingFactory };
